spring-cloud-contract-tools/spring-cloud-contract-gradle-plugin/src/main/java/org/springframework/cloud/contract/verifier/plugin/SpringCloudContractVerifierGradlePlugin.java

```diff
@@ -97,6 +107,8 @@ public void apply(Project project) {
 		createGenerateTestsTask(extension, contractTestSourceSet, copyContracts);
 		createAndConfigurePublishStubsToScmTask(extension, generateClientStubs);
 
+		project.getDependencies().add(CONTRACT_TEST_GENERATOR_RUNTIME_CLASSPATH_CONFIGURATION_NAME, "org.springframework.cloud:spring-cloud-contract-converters:" + SPRING_CLOUD_VERSION);
```

Comment:
Why do we have to add this?

Reply:
So the reason for this is because end users who apply the sc-contract plugin don't typically have the converters module on the classpath and the `RecursiveFilesConverter` lives there.

While definitely could keep this on the classpath as it is a direct dependency (would probably need to add back it's transitives), it seemed like a good idea to keep everything isolated to ensure conflicts couldn't arise via the converters module either.

Reply:
But if it's required by the plugin in general then maybe we should add it to the plugin's classpath directly?

Reply:
So since this seems to still be a little confusing (understandably), let me try to illustrate it this way.

sc-contract-converters has direct dependencies on Groovy and sc-contract-spec. Since we know the Kotlin support is fairly isolated away in sc-contract-spec-kotlin, it's not such a big deal for the Kotlin aspect, but because it is in this same area it _could_ be possible that at some point in the future Kotlin needs to be added as a direct dependency of sc-contract-spec. If that were to happen Kotlin DSL based Gradle builds would likely break again in EXACTLY the same way that they are broken today. The other facet of this relates to Groovy. While Spring Cloud version aligns with Spring Boot and furthermore Spring Framework when it comes to both Groovy and Kotlin versions, Gradle **does not** version align with any of those. So when Gradle inevitably upgrades to Groovy 3 for the Groovy DSL `build.gradle` scripts it is **likely** sc-contract will break from a Groovy standpoint instead of a Kotlin one as is the example today.

As you mentioned already, the plugin **needs** to have sc-contract-converters to perform it's job. With Gradle we declare what we need via configurations (generally dependencies). We do have options for how to **declare** those dependencies based on the **needs** that we have. The typical way for a plugin to perform it's activities is to include the necessary dependencies on its **own** implementation or api configurations as needed. If however we **need** greater isolation, we can create a configuration in the end users project model which we can either hide from them OR provide them the ability to contribute necessary dependencies to. Here we are performing the later because we need to become isolated away from Gradle itself, so that it's internal dependencies can not effect our classpath and our classpath can't conflict with Gradle's internal/parent classpath (that will contain Groovy and Kotlin + various other libraries Gradle uses).

To give this a Maven parallel, it's very similar to how plugins allow you to add additional dependencies to themselves and Maven will fork the plugin execution to an independent Java process. Same concept here.